### ISSUE TITLE
Issues connecting to restarted kernel when running jupyterhub behind Apache reverse proxy

### ISSUE THREAD

**Issue by mogthesprog** [jupyterhub/jupyterhub#422]
Hi Guys,

This is likely an issue with my Apache config, but thought it might be worth documenting it here, hopefully someone can figure out where I'm going wrong.

I've got jupytrehub running behind an apache proxy. Here is my config.

```
ProxyPass /jupyter http://localhost:8000/jupyter
ProxyPassReverse /jupyter http://localhost:8000/jupyter

Header edit Origin us-lou-svb-2006.domain/jupyter localhost:8000
RequestHeader edit Origin us-lou-svb-2006.domain/jupyter localhost:8000

Header edit Referer us-lou-svb-2006.domain/jupyter localhost:8000
RequestHeader edit Referer us-lou-svb-2006.domain/jupyter localhost:8000
<Location ~ "/(user/[^/]*)/(api/kernels/[^/]+/channels|terminals/websocket)/?">
    ProxyPass ws://localhost:8000
    ProxyPassReverse ws://localhost:8000
</Location>
```

When I restart the kernel, it looks like the kernel is restarted successfully, no errors are thrown in the web UI or in the logs. Then if i try running cells they just hang with the [*] symbol to the left of the cell

Then if I refresh the page, the kernel connects no problem.

Any ideas?

Cheers,

Morgan

**Comment by minrk** [comment]
Can you temporarily bypass Apache, and see if you see the same thing without Apache in the way?

cc @rgbkrk for more proxy knowledge

**Comment by rgbkrk** [comment]
That's an interesting one, since it sounds like your websockets are working the rest of the time. The restart action is a POST.

What browser are you on and what version of the notebook is being spawned for users?

**Comment by mogthesprog** [comment]
sorry for the slow reply, i'm going to try without the proxy in front, will post back in a while :)

**Comment by mogthesprog** [comment]
okay, so it works without the reverse proxy... So it's the reverse proxy is the problem. It seems the kernel never reports "Kernel Ready" when i'm using it with the reverse proxy. And by that i mean next to the "| python 2" at the top right, the notification reporting "Kernel Ready" never appears when i'm using the proxy, but it does when i'm going to jupyterhub directly... 

Any ideas if there's something in my config that's breaking this?

Cheers,

**Comment by minrk** [comment]
When you say that it works after you refresh the page, does that mean execution and output all work as expected?

**Comment by mogthesprog** [comment]
yep. If i restart the kernel (when using the proxy) then the kernel doesn't work. However, if i then proceed to refresh the same page, then the kernel connects and reports "Kernel Ready" and then all is well :)

**Comment by shantanudev** [comment]
I am also facing the same issue using nginx

**Comment by vroetman** [comment]
I don't think ProxyPass works with Location ~ (LocationMatch).
I did this, and it seems to work for the most part, but if I restart numerous times, it sometimes loses the connection to the websocket.

    ProxyPassMatch "^/jupyterhub/user/([^/]+)/api/kernels/([^/]+)/channels(.*)$" \
        ws://localhost:8888/jupyterhub/user/$1/api/kernels/$2/channels$3
    ProxyPassMatch "^/jupyterhub/user/([^/]+)/terminals/websocket/(.+)$" \
        ws://localhost:8888/jupyterhub/user/$1/terminals/websocket/$2
    ProxyPass        /jupyterhub  http://localhost:8888/notebook
    ProxyPassReverse /jupyterhub  http://localhost:8888/notebook

**Comment by warrenson** [comment]
I am having the same issue when changing the kernel in a running notebook to a __bash__ kernel. Changing to a __python3__ kernel works fine.

Both apache config solutions, with and without using LocationMatch, have the same symptoms which @mogthesprog initially reported.

Again, when refreshing the notebook page the bash kernel connected successfully.

So perhaps my issue is specific to the [bash kernel](https://github.com/takluyver/bash_kernel)?

Has anyone else experienced this issue for a certain kernel specifically?

**Comment by EdwardJKim** [comment]
I had the exact same issue as @mogthesprog described. It seems to be related to the [bug in Apache with secure websockets](https://bz.apache.org/bugzilla/show_bug.cgi?id=55320). The bug was fixed in 2.4.10, but I was running 2.4.6 on CentOS 7, so I rebuilt Apache from source. Dockerfile is [here](https://github.com/edwardjkim/httpd-docker) if anyone is interested. If you are running Apache 2.2, you can [backport the patch](https://gist.github.com/vitkin/6661683).

**Comment by willingc** [comment]
Thanks @EdwardJKim for passing along the information. :cake:

**Issue Resolution**
Options: 1) Run Apache 2.4.10 or higher. 2) Backport the patch to Apache 2.2. 3) Use Docker and Dockerfile.
